libbeat/autodiscover/providers/kubernetes/kubernetes.go

```diff
@@ -154,3 +195,80 @@ func (p *Provider) publish(event bus.Event) {
 	p.appenders.Append(event)
 	p.bus.Publish(event)
 }
+
+func (p *Provider) startLeading(uuid string, eventID string) {
+	event := bus.Event{
+		"start":    true,
+		"provider": uuid,
+		"id":       eventID,
+		"unique":   "true",
+	}
+	if config := p.templates.GetConfig(event); config != nil {
+		event["config"] = config
+	}
+	p.bus.Publish(event)
+}
+
+func (p *Provider) stopLeading(uuid string, eventID string) {
+	event := bus.Event{
+		"stop":     true,
+		"provider": uuid,
+		"id":       eventID,
+		"unique":   "true",
+	}
+	if config := p.templates.GetConfig(event); config != nil {
+		event["config"] = config
+	}
+	p.bus.Publish(event)
+}
+
+func (p *Provider) initLeaderElectionConfig(client k8s.Interface, uuid string) {
+	var id string
+	if p.config.Node != "" {
+		id = "beats-leader-" + p.config.Node
+	} else {
+		id = "beats-leader-" + uuid
+	}
+	lease := metav1.ObjectMeta{
+		Name:      p.config.LeaderLease,
+		Namespace: "default",
+	}
+	metaUID := lease.GetObjectMeta().GetUID()
+	p.leaderElection = leaderelection.LeaderElectionConfig{
+		Lock: &resourcelock.LeaseLock{
+			LeaseMeta: lease,
+			Client:    client.CoordinationV1(),
+			LockConfig: resourcelock.ResourceLockConfig{
+				Identity: id,
+			},
+		},
+		ReleaseOnCancel: true,
+		LeaseDuration:   15 * time.Second,
+		RenewDeadline:   10 * time.Second,
+		RetryPeriod:     2 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				p.logger.Debugf("leader election lock GAINED, id %v", id)
+				eventID := fmt.Sprintf("%v-%v", metaUID, time.Now().UnixNano())
+				p.startLeading(uuid, eventID)
+			},
+			OnStoppedLeading: func() {
+				p.logger.Debugf("leader election lock LOST, id %v", id)
+				eventID := fmt.Sprintf("%v-%v", metaUID, time.Now().UnixNano())
+				p.stopLeading(uuid, eventID)
+			},
+		},
+	}
+}
+
+func initUniqueTemplate(config *Config) {
+	m := make(map[string]interface{})
+	m["unique"] = "true"
```

Comment:
Is this `unique` condition needed? We are sending start and stop events when we start or stop leading.

Then I guess that the `unique` fields we are adding on these events wouldn't be needed either.

Reply:
The condition is needed in order to match the event and get the proper config. Otherwise, I think we the config template will "match" every event and then we will have the cluster wide config to be spawned for every autodiscover event?

Reply:
Oh, you are actually right, there will be more events.

This makes me wonder two things:
* Does it make sense to listen to other events when the provider is configured with `unique: true`? On this scenario we would only be interested on the events we provoke on `startLeading` and `stopLeading`. With current implementation other events won't produce any config because they won't have the `unique` field. We could start the provider without eventers when configured with `unique: true`.
* Should we support other conditions when the provider is configured with `unique: true`? With current implementation it would be possible to add templates with conditions, but here we are replacing them. We could support conditions if here we create the condition as `unique: true AND configured_condition`. Something like:
```
		template.ConditionConfig = &conditions.Config{
			And: []conditions.Config{
				&conditions.Config{Contains: fields},
				template.ConditionConfig,
			},
		}
```

Reply:
About point 1, I agree we can disable eventers and  just handle events coming from `startLeading` and `stopLeading`.

About point 2, based on point 1 we should remove this extra "hardcoded" condition and hence it should be doable to have configurable conditions for the templates. However I'm not sure if this would be actually useful since the events we emit are really "static" and hence I don't see any use case where conditions could be useful.

Reply:
You are right, I also cannot think on a use for conditions here.